File: backend/src/main/kotlin/com/musicstreaming/models/PlaybackState.kt
package com.musicstreaming.models

import kotlinx.serialization.Serializable

@Serializable
data class PlaybackState(
    val currentTrack: Track?,
    val positionSeconds: Long,
    val isPlaying: Boolean,
    val queue: List<Track>,
    val shuffleEnabled: Boolean,
    val timestamp: Long,
    /**
     * Direct audio stream URL for the currently playing track, if available.
     * This is typically extracted via yt-dlp by the provider (e.g. YouTube).
     * Clients should treat this as opaque and feed it directly to their audio player.
     */
    val streamUrl: String? = null,
    /**
     * Audio container/format for the current stream (e.g. WEBM, AAC).
     * Useful for clients that need to adjust playback configuration based on format.
     */
    val streamFormat: AudioFormat? = null
) {
    init {
        require(positionSeconds >= 0) { "Position must be non-negative, got: $positionSeconds" }
        if (currentTrack != null) {
            require(positionSeconds <= currentTrack.durationSeconds) {
                "Position ($positionSeconds) cannot exceed track duration (${currentTrack.durationSeconds})"
            }
        }
    }
}

File: backend-node/src/models/PlaybackState.ts
import { AudioFormat } from './StreamInfo';
import { Track } from './Track';

export interface PlaybackState {
  currentTrack: Track | null;
  positionSeconds: number;
  isPlaying: boolean;
  queue: Track[];
  shuffleEnabled: boolean;
  timestamp: number;
  streamUrl?: string | null;
  streamFormat?: AudioFormat | null;
}

export function createEmptyState(): PlaybackState {
  return {
    currentTrack: null,
    positionSeconds: 0,
    isPlaying: false,
    queue: [],
    shuffleEnabled: false,
    timestamp: Date.now(),
    streamUrl: null,
    streamFormat: null,
  };
}
